File: frontend/lib/services/food_service.dart
import '../models/food_record.dart';
import 'api_service.dart';

class FoodService {
  final ApiService _apiService = ApiService();

  /// 创建食物记录
  Future<FoodRecord> createRecord({
    required int userId,
    String photoUrl = '',
    required String foodName,
    required double calories,
    double protein = 0,
    double carbohydrates = 0,
    double fat = 0,
    double fiber = 0,
    double portion = 0,
    String unit = '',
    required String mealType,
    DateTime? eatenAt,
  }) async {
    final data = {
      'user_id': userId,
      'photo_url': photoUrl,
      'food_name': foodName,
      'calories': calories,
      'protein': protein,
      'carbohydrates': carbohydrates,
      'fat': fat,
      'fiber': fiber,
      'portion': portion,
      'unit': unit,
      'meal_type': mealType,
      if (eatenAt != null) 'eaten_at': eatenAt.toIso8601String(),
    };

    final response = await _apiService.post('/food/record', data);
    return FoodRecord.fromJson(response.data);
  }

  /// 获取食物记录列表
  Future<List<FoodRecord>> getRecords({
    required int userId,
    DateTime? startDate,
    DateTime? endDate,
  }) async {
    final queryParameters = <String, dynamic>{
      'user_id': userId,
    };

    if (startDate != null) {
      queryParameters['start_date'] = _formatDate(startDate);
    }
    if (endDate != null) {
      queryParameters['end_date'] = _formatDate(endDate);
    }

    final response = await _apiService.get(
      '/food/records',
      queryParameters: queryParameters,
    );

    final dataList = response.data['records'] as List;
    return dataList.map((item) => FoodRecord.fromJson(item)).toList();
  }

  /// 获取每日营养汇总
  Future<DailyFoodSummary> getDailySummary({
    required int userId,
    DateTime? date,
  }) async {
    final queryParameters = <String, dynamic>{
      'user_id': userId,
    };

    if (date != null) {
      queryParameters['date'] = _formatDate(date);
    }

    final response = await _apiService.get(
      '/food/daily-summary',
      queryParameters: queryParameters,
    );

    return DailyFoodSummary.fromJson(response.data);
  }

  /// 获取单个食物记录
  Future<FoodRecord> getRecord(int recordId) async {
    final response = await _apiService.get('/food/record/$recordId');
    return FoodRecord.fromJson(response.data);
  }

  /// 更新食物记录
  Future<FoodRecord> updateRecord(int recordId, {
    String? foodName,
    double? calories,
    double? protein,
    double? carbohydrates,
    double? fat,
    double? fiber,
    String? mealType,
    DateTime? eatenAt,
  }) async {
    final data = <String, dynamic>{};

    if (foodName != null) data['food_name'] = foodName;
    if (calories != null) data['calories'] = calories;
    if (protein != null) data['protein'] = protein;
    if (carbohydrates != null) data['carbohydrates'] = carbohydrates;
    if (fat != null) data['fat'] = fat;
    if (fiber != null) data['fiber'] = fiber;
    if (mealType != null) data['meal_type'] = mealType;
    if (eatenAt != null) data['eaten_at'] = eatenAt.toIso8601String();

    final response = await _apiService.put('/food/record/$recordId', data);
    return FoodRecord.fromJson(response.data);
  }

  /// 删除食物记录
  Future<void> deleteRecord(int recordId) async {
    await _apiService.delete('/food/record/$recordId');
  }

  String _formatDate(DateTime date) {
    return '${date.year}-${date.month.toString().padLeft(2, '0')}-${date.day.toString().padLeft(2, '0')}';
  }
}

File: backend/internal/services/food_service.go
package services

import (
	"time"

	"github.com/your-org/loss-weight/backend/internal/models"
	"go.uber.org/zap"
	"gorm.io/gorm"
)

type FoodService struct {
	db     *gorm.DB
	logger *zap.Logger
}

func NewFoodService(db *gorm.DB, logger *zap.Logger) *FoodService {
	return &FoodService{
		db:     db,
		logger: logger,
	}
}

type CreateFoodRecordRequest struct {
	UserID        uint      `json:"user_id" binding:"required"`
	PhotoURL      string    `json:"photo_url"`
	FoodName      string    `json:"food_name" binding:"required"`
	Calories      float32   `json:"calories" binding:"required"`
	Protein       float32   `json:"protein"`
	Carbohydrates float32   `json:"carbohydrates"`
	Fat           float32   `json:"fat"`
	Fiber         float32   `json:"fiber"`
	Portion       float32   `json:"portion"`
	Unit          string    `json:"unit"`
	MealType      string    `json:"meal_type" binding:"required"`
	EatenAt       time.Time `json:"eaten_at"`
}

type UpdateFoodRecordRequest struct {
	FoodName      string  `json:"food_name"`
	Calories      float32 `json:"calories"`
	Protein       float32 `json:"protein"`
	Carbohydrates float32 `json:"carbohydrates"`
	Fat           float32 `json:"fat"`
	Fiber         float32 `json:"fiber"`
	MealType      string  `json:"meal_type"`
	EatenAt       time.Time `json:"eaten_at"`
}

func (s *FoodService) CreateRecord(req *CreateFoodRecordRequest) (*models.FoodRecord, error) {
	record := &models.FoodRecord{
		UserID:        req.UserID,
		PhotoURL:      req.PhotoURL,
		FoodName:      req.FoodName,
		Calories:      req.Calories,
		Protein:       req.Protein,
		Carbohydrates: req.Carbohydrates,
		Fat:           req.Fat,
		Fiber:         req.Fiber,
		Portion:       req.Portion,
		Unit:          req.Unit,
		MealType:      req.MealType,
		EatenAt:       req.EatenAt,
	}

	if req.EatenAt.IsZero() {
		record.EatenAt = time.Now()
	}

	if err := s.db.Create(record).Error; err != nil {
		s.logger.Error("failed to create food record", zap.Error(err))
		return nil, err
	}

	return record, nil
}

func (s *FoodService) GetRecordsByUser(userID uint, startDate, endDate time.Time) ([]models.FoodRecord, error) {
	var records []models.FoodRecord
	query := s.db.Where("user_id = ?", userID)
	
	if !startDate.IsZero() {
		query = query.Where("eaten_at >= ?", startDate)
	}
	if !endDate.IsZero() {
		// Half-open: endDate is exclusive. Handler converts the inclusive
		// client-supplied YYYY-MM-DD into next-day midnight (in client tz).
		query = query.Where("eaten_at < ?", endDate)
	}

	if err := query.Order("eaten_at DESC").Find(&records).Error; err != nil {
		s.logger.Error("failed to get food records", zap.Error(err))
		return nil, err
	}

	return records, nil
}

func (s *FoodService) GetRecordByID(id uint) (*models.FoodRecord, error) {
	var record models.FoodRecord
	if err := s.db.First(&record, id).Error; err != nil {
		s.logger.Error("failed to get food record", zap.Error(err))
		return nil, err
	}
	return &record, nil
}

func (s *FoodService) UpdateRecord(id uint, req *UpdateFoodRecordRequest) (*models.FoodRecord, error) {
	var record models.FoodRecord
	if err := s.db.First(&record, id).Error; err != nil {
		s.logger.Error("failed to get food record", zap.Error(err))
		return nil, err
	}

	updates := make(map[string]interface{})
	if req.FoodName != "" {
		updates["food_name"] = req.FoodName
	}
	if req.Calories > 0 {
		updates["calories"] = req.Calories
	}
	if req.Protein > 0 {
		updates["protein"] = req.Protein
	}
	if req.Carbohydrates > 0 {
		updates["carbohydrates"] = req.Carbohydrates
	}
	if req.Fat > 0 {
		updates["fat"] = req.Fat
	}
	if req.Fiber > 0 {
		updates["fiber"] = req.Fiber
	}
	if req.MealType != "" {
		updates["meal_type"] = req.MealType
	}
	if !req.EatenAt.IsZero() {
		updates["eaten_at"] = req.EatenAt
	}

	if err := s.db.Model(&record).Updates(updates).Error; err != nil {
		s.logger.Error("failed to update food record", zap.Error(err))
		return nil, err
	}

	return &record, nil
}

func (s *FoodService) DeleteRecord(id uint) error {
	if err := s.db.Delete(&models.FoodRecord{}, id).Error; err != nil {
		s.logger.Error("failed to delete food record", zap.Error(err))
		return err
	}
	return nil
}

func (s *FoodService) GetDailySummary(userID uint, date time.Time, loc *time.Location) (map[string]interface{}, error) {
	startOfDay := StartOfDay(date, loc)
	endOfDay := startOfDay.Add(24 * time.Hour)

	var records []models.FoodRecord
	if err := s.db.Where("user_id = ? AND eaten_at >= ? AND eaten_at < ?", userID, startOfDay, endOfDay).
		Find(&records).Error; err != nil {
		s.logger.Error("failed to get daily food records", zap.Error(err))
		return nil, err
	}

	var totalCalories, totalProtein, totalCarbs, totalFat float32
	for _, record := range records {
		totalCalories += record.Calories
		totalProtein += record.Protein
		totalCarbs += record.Carbohydrates
		totalFat += record.Fat
	}

	return map[string]interface{}{
		"date":            date,
		"total_calories":  totalCalories,
		"total_protein":   totalProtein,
		"total_carbs":     totalCarbs,
		"total_fat":       totalFat,
		"meal_count":      len(records),
		"records":         records,
	}, nil
}
